File: baekjoon/스텍/오큰수.py
n = int(input())
arr = list(map(int,input().split()))

answer = []
for i in range(0,n-1):
    nge = arr[i]
    for j in range(i+1,n):
        if arr[j] > nge:
            nge = arr[j]
            break

    if arr[i] == nge:
        answer.append(-1)
    else:
        answer.append(nge)
    
answer.append(-1)

for i in answer:
    print(i,end=" ")

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int[] answer = new int[n];
        int[] numbers = new int[n];

        String[] input = br.readLine().split(" ");
        for (int i = 0; i < n; i++) {
            numbers[i] = Integer.parseInt(input[i]);
        }

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            if (stack.isEmpty()) {
                stack.push(i);
                continue;
            }
            int size = stack.size();
            for (int j = 0; j < size; j++) {
                if (numbers[stack.peek()] >= numbers[i]) {
                    break;
                }

                answer[stack.pop()] = numbers[i];
            }
            stack.push(i);
        }

        int size = stack.size();
        for (int i = 0; i < size; i++) {
            answer[stack.pop()] = -1;
        }

        StringBuilder sb = new StringBuilder();
        for (int a : answer) {
            sb.append(a).append(" ");
        }

        System.out.println(sb.toString());
    }
}
